# exploratory data analysis - bidding zone NO1 and neighboring zones for the year of 2022

In [1]:
# importing dependencies to aggregate dataset
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

**About:**
Hourly electricity prices and external factors for bidding zone NO1 (Norway Østlandet) and neightbouring zones (NO2, NO3, NO5 and SE3).
**Source of the datasets data::** https://transparency.entsoe.eu/dashboard/show

**Attribute information:**

**Target feature:** 	**Day-ahead Price** Spot (day-ahead) market prices in the currency Euro per MWh (megawatt-hour).
1.	**Start MTU (UTC):** Start datetime of the interval for the measured values
2.	**Start MTU (UTC):** End time of the interval for the measured values
3.	**Actual Total Load:** The power consumption in each zone per MW (megawatt)
4.	**Biomass:** Energy production in MW  from Biomass.
5.	**Fossil Gas:** Energy production in MW  from Fossil Gas.
6.	**Hydro Run-of-river and poundage:** Energy production in MW  from Hydro Run-of-river and poundage.
7.	**Hydro Water Reservoir:** Energy production in MW  from Hydro Run-of-river and poundage.
8.	**Wind Onshore:** Energy production in MW  from onshore wind.
9.	**Hydro Pumped Storage Aggregated:** Energy production in MW  from Hydro Pumped Storage Aggregated.
10.	**Hydro Run-of-river and poundage:** Energy production in MW  from Hydro Run-of-river and poundage
11.	**Other renewable:** Energy production in MW  from Other renewable.
12.	**Marine:** Energy production in MW  from Marine.
13.	**Solar:** Energy production in MW  from Solar.
14.	**CBF BZN|x > BZN|y [MWh]:**  Cross-border physical flow in MWH (mega watt hour), zone x is export zone and zone y is import zone.
15.	**Stored Energy Value Water Reservoirs and Hydro Storage Plants [MWh]:** Stored Energy Value in Water Reservoirs and Hydro Storage Plants measured in MwH (mega watt hour)

In [2]:
# reding in the dataset to a dataframe
no1 = pd.read_csv('../datasets/no1_aggregated/no1_2020_to_2022_aggregated.csv', index_col=0)

# printing out he 5 first rows of the dataset
no1.head()

start MTU (UTC)        end MTU (UTC)  \
0  2020-01-01 00:00:00  2020-01-01 01:00:00   
1  2020-01-01 01:00:00  2020-01-01 02:00:00   
2  2020-01-01 02:00:00  2020-01-01 03:00:00   
3  2020-01-01 03:00:00  2020-01-01 04:00:00   
4  2020-01-01 04:00:00  2020-01-01 05:00:00   

   Day-ahead Price [EUR/MWh] BZN|NO1  Day-ahead Price [EUR/MWh] BZN|NO2  \
0                              31.77                              31.77   
1                              31.57                              31.57   
2                              31.28                              31.28   
3                              30.72                              30.72   
4                              30.27                              30.27   

   Day-ahead Price [EUR/MWh] BZN|NO3  Day-ahead Price [EUR/MWh] BZN|NO5  \
0                              28.45                              31.77   
1                              27.90                              31.57   
2                              27.52                              31.28   
3                              27.54                              30.72   
4                              26.55                              30.27   

   Day-ahead Price [EUR/MWh] BZN|SE3  Actual Total Load [MW] - BZN|NO1  \
0                              28.45                            4333.0   
1                              27.90                            4250.0   
2                              27.52                            4167.0   
3                              27.54                            4145.0   
4                              26.55                            4222.0   

   Actual Total Load [MW] - BZN|NO2  Actual Total Load [MW] - BZN|NO3  ...  \
0                            4139.0                            3016.0  ...   
1                            4114.0                            2945.0  ...   
2                            4030.0                            3032.0  ...   
3                            4032.0                            2988.0  ...   
4                            4032.0                            2848.0  ...   

   CBF BZN|NO1 > BZN|NO3 [MW]  CBF BZN|NO5 > BZN|NO1 [MW]  \
0                         0.0                      1449.0   
1                         0.0                      1283.0   
2                         6.0                       962.0   
3                         6.0                       767.0   
4                        22.0                       642.0   

   CBF BZN|NO1 > BZN|NO5 [MW]  CBF BZN|SE3 > BZN|NO1 [MW]  \
0                         0.0                         391   
1                         0.0                         348   
2                         0.0                         500   
3                         0.0                         610   
4                         0.0                         909   

   CBF BZN|NO1 > BZN|SE3 [MW]  \
0                           0   
1                           0   
2                           0   
3                           0   
4                           0   

   Stored Energy Value Water Reservoirs and Hydro Storage Plants [MWh] - BZN|NO1  \
0                                       3.962493e+06                               
1                                       3.961448e+06                               
2                                       3.960404e+06                               
3                                       3.959359e+06                               
4                                       3.958314e+06                               

   Stored Energy Value Water Reservoirs and Hydro Storage Plants [MWh] - BZN|NO2  \
0                                       2.438936e+07                               
1                                       2.439155e+07                               
2                                       2.439372e+07                               
3                                       2.439589e+07                               
4                                       2.439805e+

In [3]:
# printing out the shape of the dataset
no1.shape

(26304, 57)

In [4]:
# printing out info about dataset
no1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26304 entries, 0 to 26303
Data columns (total 57 columns):
 #   Column                                                                         Non-Null Count  Dtype  
---  ------                                                                         --------------  -----  
 0   start MTU (UTC)                                                                26304 non-null  object 
 1   end MTU (UTC)                                                                  26304 non-null  object 
 2   Day-ahead Price [EUR/MWh] BZN|NO1                                              26304 non-null  float64
 3   Day-ahead Price [EUR/MWh] BZN|NO2                                              26304 non-null  float64
 4   Day-ahead Price [EUR/MWh] BZN|NO3                                              26304 non-null  float64
 5   Day-ahead Price [EUR/MWh] BZN|NO5                                              26304 non-null  float64
 6   Day-ahead Price [EUR/M

We Can see that the datetime columns are of the data type object, so we ned to convert it into the data type datetime. The remaining values seems to have an appropriate data type. We also notice that the value onthe 'non-null count' of allot of the columns are less than the size of the dataset (26304), and we therefore know that allot of the features have missing values. We should therefore investigate which values are misssing, and handle the missing values accordingly.

In [5]:
# converting datetime columns to data type datetime
no1['start MTU (UTC)'] = pd.to_datetime(no1['start MTU (UTC)'])
no1['start MTU (UTC)'] = pd.to_datetime(no1['start MTU (UTC)'])

In [6]:
# printing out summary of missing values in the dataset:
no1.isna().sum()

start MTU (UTC)                                                                   0
end MTU (UTC)                                                                     0
Day-ahead Price [EUR/MWh] BZN|NO1                                                 0
Day-ahead Price [EUR/MWh] BZN|NO2                                                 0
Day-ahead Price [EUR/MWh] BZN|NO3                                                 0
Day-ahead Price [EUR/MWh] BZN|NO5                                                 0
Day-ahead Price [EUR/MWh] BZN|SE3                                                 0
Actual Total Load [MW] - BZN|NO1                                                  1
Actual Total Load [MW] - BZN|NO2                                                  1
Actual Total Load [MW] - BZN|NO3                                                  1
Actual Total Load [MW] - BZN|NO5                                                  1
Actual Total Load [MW] - BZN|SE3                                            

We can now see which features are missing values, and how many values are missing from each feature. We will handle the missing values feature by feature starting with 'Actual Total Load [MW]'

We can start by checking which rows are missing values for the load attributes:

In [19]:
# creating a list of load column names
load_column_names = [col_name for col_name in no1.columns if 'Actual Total Load' in col_name]
print(load_column_names)
                    

# checking which row is missing values for column 'Actual Total Load [MW] - BZN|NO1'
no1[no1['Actual Total Load [MW] - BZN|NO1'].isnull()].loc[:, load_column_names]

['Actual Total Load [MW] - BZN|NO1', 'Actual Total Load [MW] - BZN|NO2', 'Actual Total Load [MW] - BZN|NO3', 'Actual Total Load [MW] - BZN|NO5', 'Actual Total Load [MW] - BZN|SE3']


Actual Total Load [MW] - BZN|NO1  Actual Total Load [MW] - BZN|NO2  \
16057                               NaN                               NaN   

       Actual Total Load [MW] - BZN|NO3  Actual Total Load [MW] - BZN|NO5  \
16057                               NaN                               NaN   

       Actual Total Load [MW] - BZN|SE3  
16057                              7786

It seems like none of the Norwegian bidding zones have recorded data for the time interval 31.10.2021 01:00 - 31.10.2021 02:00.

Since it is a dingle row with a missing vlaue, we can simply impute the missing value through interpolation which uses the neighbouring values to generate an the missing intermediary value. We choose to use spline interpolation which estimates the missing values while focusing on minimizing the curvature  of the value graph. 

In [20]:
# performing splilne inerpolation to fill missing values in load columns
no1['Actual Total Load [MW] - BZN|NO1'] = no1['Actual Total Load [MW] - BZN|NO1'].interpolate(option='spline')
no1['Actual Total Load [MW] - BZN|NO2'] = no1['Actual Total Load [MW] - BZN|NO2'].interpolate(option='spline')
no1['Actual Total Load [MW] - BZN|NO3'] = no1['Actual Total Load [MW] - BZN|NO3'].interpolate(option='spline')
no1['Actual Total Load [MW] - BZN|NO5'] = no1['Actual Total Load [MW] - BZN|NO5'].interpolate(option='spline')
no1['Actual Total Load [MW] - BZN|SE3'] = no1['Actual Total Load [MW] - BZN|SE3'].interpolate(option='spline')

# checking if all the missing values in load columns have been filled:
no1.loc[:, load_column_names].isna().sum()

Actual Total Load [MW] - BZN|NO1    0
Actual Total Load [MW] - BZN|NO2    0
Actual Total Load [MW] - BZN|NO3    0
Actual Total Load [MW] - BZN|NO5    0
Actual Total Load [MW] - BZN|SE3    0
dtype: int64

We can see that none of the load columns have any missing values